qt5: fix cross

amjoseph-nixpkgs opened the pull request:
Reengineered to avoid rebuilding anything.  Should have ~0 rebuilds.

To be useful, you also need this two-commit PR from staging (which *does* cause rebuilds; that's why it's a separate PR):

- https://github.com/NixOS/nixpkgs/pull/264964
 
## Description of changes

This commit fixes builds of `pkgsCross.*.qt5.qtbase` by:

- Adding the buildPlatform compiler to depsBuildBuild in qtbase.nix  and qtModule.nix.  The `qtbase` build machinery expects to find it  in the $PATH in unprefixed form.

- Setting the `PKG_CONFIG_SYSROOT_DIR` and `PKG_CONFIG_LIBDIR`  environment variables when compiling a cross-targeted `qmake`.  This is required; if these environment variables are unset,  `qmake` won't even try to use `pkg-config`.

- Adding the `-device` and `-device-option` flags necessary for  cross compilation to `configureFlags`.

- Adding the (one-entry at the moment) Rosetta Stone for QT-5 as a  `let`-defined `qtPlatform` function which takes a nixpkgs platform  and returns a QT-5 `mkspecs`-string.

Includes

- https://github.com/NixOS/nixpkgs/pull/264903
- https://github.com/NixOS/nixpkgs/pull/265134


<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Commit:
lib.systems.inspect: add patternLogicalAnd

Artturin commented on pkgs/development/python-modules/pyqtwebengine/default.nix:
```diff
@@ -44,8 +46,17 @@ in buildPythonPackage rec {
+    qtdeclarative
   ];
 
+  # TODO: figure out why the env hooks aren't adding these inclusions automatically
```
Only the `/include` is added , not any sub paths IIRC

Many of the includes in https://github.com/search?q=%2F%23Include.%2BQtWebChannel%2F&type=code have a / so I don't know if they're supposed to be added to it.

Commit:
qutebrowser: repair vandalism

Our qutebrowser package has been vandalized by various commits, such
as c9cc3a2e3eb4661b49176e0c62afd097793c9626 and
ad0bbaf1d21f3b5da9c9cf68f0583d487c295782, which made erroneous
assertions such as "since qutebrowser 3.0.0 the derivation is only
building for qt6."

This commit repairs the vandalism.

amjoseph-nixpkgs commented:
@ofborg build pkgsCross.aarch64-multiplatform.qt5.qutebrowser

amjoseph-nixpkgs commented:
@ofborg build pkgsCross.aarch64-multiplatform.qt5.qutebrowser

amjoseph-nixpkgs commented:
@ofborg build pkgsCross.aarch64-multiplatform.qt5.qutebrowser

Artturin commented on pkgs/applications/networking/browsers/qutebrowser/default.nix:
```diff
@@ -12,6 +12,7 @@
 , widevine-cdm
 , enableVulkan ? stdenv.isLinux
 , vulkan-loader
+, isQt6 ? false
```
Could do it automatically with something like

`isQt6 = lib.versions.major qtbase.version == "6";`
zeal and memento have that for qt5

amjoseph-nixpkgs replied:
c3458dd757c1

amjoseph-nixpkgs commented:
@ofborg build pkgsCross.aarch64-multiplatform.qt5.qutebrowser

Artturin commented on pkgs/development/python-modules/pyqtwebengine/default.nix:
```diff
@@ -21,20 +22,21 @@ in buildPythonPackage rec {
     pkg-config
     qmake
-    sip
+    buildPackages.python3Packages.sip
```
Let's see what the consequences of passing spliced as self in makeScopeWithSplicing arehttps://github.com/NixOS/nixpkgs/pull/265136

amjoseph-nixpkgs replied:
I don't understand this comment.

Aren't we trying to move away from using unspliced packages in dependency attributes like `nativeBuildInputs`?

Artturin replied:
The change you have here now is necessary because of https://github.com/NixOS/nixpkgs/blob/ad83c5cf4b7b05b134e9698ff3f30069829d7fa6/pkgs/top-level/python-packages.nix#L10730-L10732 `self` doesn't contain spliced packages

amjoseph-nixpkgs replied:
:rage: :rage: :rage:  I hate splicing. :rage: :rage: :rage:

Artturin commented on pkgs/development/libraries/qt-5/modules/qtbase.nix:
```diff
@@ -381,4 +410,12 @@ stdenv.mkDerivation (finalAttrs: {
   };
 
+}))
+.overrideAttrs(previousAttrs: lib.optionalAttrs (stdenv.buildPlatform != stdenv.hostPlatform) {
+  configurePlatforms = [ ];
+
+  # `qtbase` expects to find `cc` (with no prefix) in the
+  # `$PATH`, so the following is needed even if
+  # `stdenv.buildPlatform.canExecute stdenv.hostPlatform`
+  depsBuildBuild = (previousAttrs.depsBuildBuild or []) ++ [ buildPackages.stdenv.cc ];
 })
```
Shouldn't need to be an overrideAttrs, a few `optionals` should work

amjoseph-nixpkgs replied:
There is no `depsBuildBuild` in the original expression, so adding `depsBuildBuild=[]` to the native side will trigger rebuilds.

I expect that `staging` will merge to `master` before this PR is merged.  If that happens and I can avoid rebuilds, I will change the branch of this PR from `staging` to `master`.

amjoseph-nixpkgs replied:
This PR now causes no rebuilds.

Artturin commented on pkgs/development/libraries/qt-5/5.15/default.nix:
```diff
@@ -356,5 +356,8 @@ let
   finalScope = baseScope.overrideScope(final: prev: {
     qttranslations = bootstrapScope.qttranslations;
+    qutebrowser = final.callPackage ../../../../applications/networking/browsers/qutebrowser {
+      inherit (final) qtbase qtwebengine wrapQtAppsHook qtwayland;
```
should be unnecessary to inherit

amjoseph-nixpkgs replied:
11bb9d3d3b50

amjoseph-nixpkgs commented:
@ofborg build pkgsCross.aarch64-multiplatform.qt5.qutebrowser

amjoseph-nixpkgs commented:
@ofborg build pkgsCross.aarch64-multiplatform.qt5.qutebrowser

Commit:
qt5.qmake-hook: move libs to depsTargetTargetPropagated

Commit:
qt5.qtdeclarative: add postFixup if cross compiling

Commit:
qt5.qtModule: reformat arguments

Commit:
qt5.qtModule: add buildPackages.stdenv.cc to depsBuildBuild if cross compiling

Commit:
qt5.qtModule: add explicit pkgsHostTarget.qt5.qtbase.dev to nativeBuildInputs if cross compiling

Commit:
qt5.qtwebchannel: omit "bin" output when cross compiling

Commit:
qt5.qtwebengine: fix cross

Commit:
qt5.qttranslations: disable if cross to prevent infinite recursion

Commit:
qt5.qtbase: fix cross

Commit:
qt5.pyqtwebengine: fix cross

Commit:
pyqt5: fix cross

Commit:
jasper: mark broken if cross

Commit:
qt5.qtimageformats: do not try to use broken libraries

Commit:
test.cross.sanity: add qt5.qutebrowser, firefox

Commit:
qutebrowser: take python3 from buildPackages

amjoseph-nixpkgs commented:
@ofborg eval

dotlambda commented on pkgs/applications/networking/browsers/qutebrowser/default.nix:
```diff
@@ -80,7 +86,7 @@ python3.pkgs.buildPythonApplication {
 
     make -f misc/Makefile \
-      PYTHON=${python3}/bin/python3 \
+      PYTHON=${buildPackages.python3}/bin/python3 \
```
```suggestion
      PYTHON=${python3.pythonForBuild.interpreter} \
```

amjoseph-nixpkgs commented:
This works; along with https://github.com/NixOS/nixpkgs/pull/264964:

- pkgsCross.aarch64-multiplatform.qt5.qutebrowser on x86_64-linux — [**Success**](https://github.com/NixOS/nixpkgs/pull/265401/checks?check_run_id=18368468301)

It causes no rebuilds:

![no-rebuilds](https://github.com/NixOS/nixpkgs/assets/54836058/3204ff5d-1171-4f11-bda1-92db473d7ad5)

There are a lot of conditionals; most can be removed, but doing so causes rebuilds, so the conditional-removals must go to staging.

I'm incredibly tired of rebasing this.

I'm merging it.

Maybe somebody will rage-revert it, maybe not.

In either case I'll stay on the branch (local if necessary) that keeps it in-tree, because it's kind of insane to use a distribution with a browser-only forge that repeatedly breaks your web browser.  If I can stop spending all my time rebasing this I might actually be able to finish getting qt6 to cross-compile after 23.11:

- https://github.com/NixOS/nixpkgs/pull/265524
- https://github.com/NixOS/nixpkgs/pull/264962
- https://github.com/NixOS/nixpkgs/pull/264961
- https://github.com/NixOS/nixpkgs/pull/264960
- https://github.com/NixOS/nixpkgs/pull/265523
- https://github.com/NixOS/nixpkgs/pull/265525
- https://github.com/NixOS/nixpkgs/pull/265528

dotlambda commented:
Why did you neither implement my suggestion nor wait for other reviews?

winterqt commented:
> Maybe somebody will rage-revert it, maybe not.

This is possibly the most self-aware yet utterly absurd comment I've seen here to date.

A gentle reminder that we are a team, and you are expected to work with others to meet our shared goals; completely blindsiding the active reviewers goes against that entirely.

Artturin commented on pkgs/development/libraries/qt-5/modules/qtdeclarative.nix:
```diff
@@ -21,4 +23,10 @@ qtModule {
     "bin/qmlscene"
     "bin/qmltestrunner"
   ];
+  postFixup = lib.optionalString (stdenv.buildPlatform != stdenv.hostPlatform) ''
+    mv $dev/bin/qmlformat $bin/bin/qmlformat
+    mv $dev/bin/qmltyperegistrar $bin/bin/qmltyperegistrar
+    ln -s $bin/bin/qmlformat $dev/bin/qmlformat
+    ln -s $bin/bin/qmltyperegistrar $dev/bin/qmltyperegistrar
+  '';
```
why

amjoseph-nixpkgs replied:
I'm running a build to get the exact failure, but IIRC it's related to the fact that we can't keep hostPlatform binaries out of the $PATH because:
- QT puts `qmake` and `libQtBase` in the same package
- QT goes bananas with `propagatedBuildInputs` and there is no way to "unpropagate" or "block propagation" of dependencies at a certain point in the tree.

amjoseph-nixpkgs replied:
I was mistaken; it is not necessary anymore.  At one point in my development process it was, but it no longer is.

- https://github.com/NixOS/nixpkgs/pull/265583

Artturin commented on pkgs/development/python-modules/pyqtwebengine/default.nix:
```diff
@@ -29,12 +30,18 @@ in buildPythonPackage rec {
+  ] ++ lib.optionals (stdenv.buildPlatform == stdenv.hostPlatform) [
     sip
+  ] ++ lib.optionals (stdenv.buildPlatform != stdenv.hostPlatform) [
+    buildPackages.python3Packages.sip
```
this will not work when cross-compiling a non-default python version

needs to use `python.pythonForBuild.pkgs.sip`

amjoseph-nixpkgs replied:
> needs to use `python.pythonForBuild.pkgs.sip`

`python.pythonForBuild` is `pkgsBuildBuild.python`, but we want `pkgsBuildHost.python`.  I will try to find a way to get that via `python.*`.

Artturin replied:
> > needs to use `python.pythonForBuild.pkgs.sip`
> 
> `python.pythonForBuild` is `pkgsBuildBuild.python`, but we want `pkgsBuildHost.python`. I will try to find a way to get that via `python.*`.

Sip does c stuff so I guess target matters for it?

amjoseph-nixpkgs replied:
https://github.com/NixOS/nixpkgs/pull/265570

amjoseph-nixpkgs replied:
> Sip does c stuff so I guess target matters for it?

Quite likely.  Or even using `pkg-config` would make it target-sensitive.

Artturin commented on pkgs/development/libraries/qt-5/qtModule.nix:
```diff
@@ -22,7 +23,12 @@ mkDerivation (args // {
+    (args.nativeBuildInputs or []) ++ [
+      perl qmake
+    ] ++ lib.optionals (stdenv.buildPlatform != stdenv.hostPlatform) [
+      pkgsHostTarget.qt5.qtbase.dev
```
why, the packages there shouldn't be executable?

amjoseph-nixpkgs replied:
Normally no, except it's QT, and QT has both `qmake`-which-runs-on-the-buildPlatform as well as `libQtBase`-which-is-built-for-the-hostPlatform mashed into the same package.

Artturin commented on pkgs/development/python-modules/pyqt/5.x.nix:
```diff
@@ -90,11 +137,13 @@ buildPythonPackage rec {
 
   buildInputs = with libsForQt5; [
     dbus
+  ] ++ lib.optionals (stdenv.buildPlatform == stdenv.hostPlatform) [
```
please clean up the conditionals on staging

amjoseph-nixpkgs replied:
https://github.com/NixOS/nixpkgs/pull/265573

Artturin commented on pkgs/development/libraries/qt-5/5.15/default.nix:
```diff
@@ -313,7 +313,9 @@ let
       qmake = callPackage ({ qtbase }: makeSetupHook {
         name = "qmake-hook";
-        propagatedBuildInputs = [ qtbase.dev ];
+        ${if stdenv.buildPlatform == stdenv.hostPlatform
+          then "propagatedBuildInputs"
+          else "depsTargetTargetPropagated"} = [ qtbase.dev ];
```
Seems wrong, the binaries in qtbase won't be in `PATH`

amjoseph-nixpkgs replied:
Remember, `qtbase` is both a build tool (`qmake`) and a library (`libQt5Base`).  The line above deals with the library.

Artturin replied:
Are you sure it deals with only the library? https://github.com/nixpkgs/nixpkgs/blob/753bf0e2933c4d66c0eb8481b60157419bae3629/pkgs/development/libraries/qt-5/hooks/qmake-hook.sh#L37

Artturin replied:
> Normally no, except it's QT, and QT has both qmake-which-runs-on-the-buildPlatform as well as libQtBase-which-is-built-for-the-hostPlatform mashed into the same package.

Gotta add comments ..

Artturin commented:
github removed all my nixpkgs notifications and unsubscribed me from all the nixpkgs PRs and issues :angry:

rnhmjoj commented:
You sneaked controversial changes (the qutebrowser stuff) in an unrelated PR and self-merged them without waiting for anyone to review. I think this behavior is borderline malicious and I seriously suggest that your commit rights be revoked.

winterqt commented:
> You sneaked controversial changes (the qutebrowser stuff) in an unrelated PR

At least the merged change didn't make it so that any usage of `qutebrowser` would be downgraded to an insecure version, like a previous version of that PR apparently did. 😕

amjoseph-nixpkgs commented:
> You sneaked controversial changes (the qutebrowser stuff) in

I think you are confused here.

It was in fact *you* who sneaked controversial changes to qutebrowser in https://github.com/NixOS/nixpkgs/pull/250171 which deleted `qt5.qutebrowser` (the only one that builds under many configurations) into a version-bump PR titled "qutebrowser: 2.5.4 -> 3.0.0".

nrdxp commented:
> It was in fact _you_ who sneaked controversial changes to qutebrowser

Who, besides you, agrees that the upgrade to qt6 was controversial? Certainly not upstream. In fact, it is often considered controversial to break from upstream packaging standards in any serious manner, which means you have the burden of proof here, not us.

In addition, I think most users were anxious for the qt6 version, see as the qt5 webengine was already severely outdated by the time qt6 support was made stable in qutebrowser.

Really its just plain subversive to sneak in changes like this when matainers don't agree with you, especially if you don't pick a share of the maintenance burden by adding yourself as a maintainer.

amjoseph-nixpkgs commented:
> Who, besides you, agrees that the upgrade to qt6 was controversial?

Please don't put words in my mouth.

Take a moment to read the PR that I linked to.  Before that PR qutebrowser supported both qt5 and qt6.  I have no objection to the addition of support for qt6.

dotlambda commented on pkgs/development/libraries/qt-5/5.15/default.nix:
```diff
@@ -338,6 +340,12 @@ let
+      if stdenv.buildPlatform == stdenv.hostPlatform
+      then bootstrapScope.qttranslations
+      else null;
+    qutebrowser = final.callPackage ../../../../applications/networking/browsers/qutebrowser { };
```
Applications don't belong here so I suggest #251671.
